server/website/website/views.py

```diff
@@ -486,6 +488,54 @@ def handle_result_files(session, files, execution_times=None):
     # Load the contents of the controller's summary file
     summary = JSONUtil.loads(files['summary'])
 
+    dbms_id = session.dbms.pk
+    udm_before = {}
+    udm_after = {}
+    if ('user_defined_metrics' not in files) and (USER_DEFINED_TARGET == session.target_objective):
+        return HttpResponse('ERROR: user defined target objective is not uploaded!')
+    # User defined metrics
+    udm = {}
+    if 'user_defined_metrics' in files:
+        udm = JSONUtil.loads(files['user_defined_metrics'])
+    if len(udm) > 0:
+        udm_target = udm['target_objective']
+        udm_not_target = udm['metrics']
+        udm_all = copy.deepcopy(udm_not_target)
+        if (udm_target is None) and (USER_DEFINED_TARGET == session.target_objective):
+            return HttpResponse('ERROR: user defined target objective is not uploaded!')
+        if udm_target is not None:
+            udm_all.update(udm_target)
+        if not target_objectives.udm_registered(dbms_id):
+            target_objectives.register_udm(dbms_id, udm_all)
+        for name, info in udm_all.items():
+            udm_name = 'udm.' + name
+            udm_before[name] = 0
+            udm_after[name] = info['value']
+            if MetricCatalog.objects.filter(dbms=session.dbms, name=udm_name).exists():
+                continue
+            udm_catalog = MetricCatalog.objects.create(dbms=session.dbms,
+                                                       name=udm_name,
+                                                       vartype=info['type'],
+                                                       scope='global',
+                                                       metric_type=MetricType.STATISTICS)
+            udm_catalog.summary = 'user defined metric, not target objective'
+            udm_catalog.save()
+        if udm_target is not None:
+            target_name = 'udm.' + list(udm_target.keys())[0]
+            pprint_name = 'udf.' + list(udm_target.keys())[0]
+            info = list(udm_target.values())[0]
+            if USER_DEFINED_TARGET != session.target_objective:
+                LOG.warning('the target objective is not user defined metric (UDM),\
+                            please disable UDM target objective in driver')
+            else:
+                udm_instance = target_objectives.get_instance(dbms_id, USER_DEFINED_TARGET)
+                if not udm_instance.is_registered():
```

Comment:
I am thinking of the logic here. According to the current implementation, udm_instance can only be registered once, and udm_instance can be registered after the first iteration. Are those the behaviors we want?
For GPR, target_objective can be changed multiples times and at any iteration. For DDPG, changing the target_objective after the first iteration will make the training data in the replay memory inconsistent, so it should not be allowed.
My suggestion is that you only register UDM at the first iteration, when status='default'. 
Roughly here: https://github.com/cmu-db/ottertune/blob/35bc53fcb7a81f6004f73bcf6878ff485e29975d/server/website/website/views.py#L720
But you will need to pull --rebase.

Reply:
You are right. We should only set the target objective at the first iteration, and return error message when it tries to reset the target objective afterwards.